### PR TITLE
Fix compiler warning and replace ALLOC/REALLOC/FREE by a less verbose…

### PR DESCRIPTION
**Pull request by mingodad**
… ones

### CODE REVIEW

**Comment by thutt** [comment]
Can I trouble you to make full change descriptions for each change that describes the change (why the change is being made, what the change does), and what testing was performed.  Being able to successfully run './scripts/precommit-tests' is mandatory, but any other testing information would be useful too.

The reason I ask for this is to help people actually undrestand how the code is changing (hopefully, always improving) by describing the mistakes that were made, and how they were resolved.   The testing section of the change description is important so that people can reproduce the results at any SHA, and to ensure that there are no regressions.

File: make/config.mk [hunk 1]
```diff
@@ -55,7 +55,7 @@ export DEBUG
 
 
 OPT	:=							\
-	$(if $(filter release,$(LRSTAR_BUILD_TYPE)),-Os,-Og)	\
+	$(if $(filter release,$(LRSTAR_BUILD_TYPE)),-Wall -Os,-Wall -Og)	\
```
Comment (thutt) [inline make/config.mk:58]: Thanks for doing this! I've been meaning to enable a higher level of diagnostics on lrstar and dfa, but had not gotten to work on it yet. 

Enabling '-Werror' is also highly desirable.

File: src/dfa/CM_Global.cpp [hunk 2]
```diff
@@ -355,51 +355,47 @@ int   get_fid (char *arg, char *dir, char *fn, char *ft)
...
-      n_errors++;
-      prt_log ("%s:%s: Allocation failure; %u bytes not available.\n\n",
-               fname, lineno, n_bytes);
-      Quit ();
-   }
-   memory_usage += n_bytes;
-   if (memory_usage > memory_max) memory_max = memory_usage;
-   return (x);
-}
+#ifdef _DEBUG
```
Comment (thutt) [inline src/dfa/CM_Global.cpp:358]: I'd prefer to not have conditional compilation.  I've seen too many cases where code rots because a branch of the conditional is never exercised.

This is the reason that there are two versions of the memory allocation routines in the code that I wrote.  The compiler & linker will not include the unused version in the executable.  If it is included, then the options are set wrong.

Paul preferred the debug version to have the file and line number printed in the debug version to help find the location where the memory failure occured.
Can that be preserved?

File: src/dfa/CM_Global.cpp [hunk 3]
```diff
@@ -355,51 +355,47 @@ int   get_fid (char *arg, char *dir, char *fn, char *ft)
...
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 
-void  frea  (char *x, int n_bytes)
-{
-   free(x);
-   memory_usage -= n_bytes;
+void frea(char** x, int size, int n) {
+    free(*x);
```
Comment (thutt) [inline src/dfa/CM_Global.cpp:396]: The code needed to generate this kind of interface will be worse than only having one level of pointer.    If you'd like to ensure that the pointer is set to NULL, then I recommend moving the implementation to a header file and making it a static inline function; that way the address not need to be taken.

File: src/dfa/CM_Global.h [hunk 4]
```diff
@@ -118,40 +118,14 @@ enum options
...
-   do {                                                           \
-      char         **p_        = reinterpret_cast<char **>(&_x);  \
-      const size_t   n_el_    = _n_elem;                          \
-      const size_t   el_size_  = sizeof(*_x);                     \
-      const size_t   n_bytes_ = (n_el_) * (el_size_);             \
-      frea(*p_, n_bytes_);                                        \
-      _x = 0;                                                     \
-   } while (0)
+#ifdef _DEBUG
+#define ALLOC(x,norg)    alloc (#x, (char**)&x, sizeof(*x), norg)
```
Comment (thutt) [inline src/dfa/CM_Global.h:122]: This is linked to the implementation  above, so I will let those comments speak for this location too.  In particular, having two versions with conditional compilation is less desirable than a more complicated macro that the compiler can efficiently use.

File: src/dfa/LG_Actions2.cpp [hunk 5]
```diff
@@ -40,8 +40,8 @@ static int*   L_tail;
 static char*  ebnfspace;
 static char*  ebnfspace_start;
 
-static int    c;
-static int    warning_off = 0;
+//static int    c;
```
Comment (thutt) [inline src/dfa/LG_Actions2.cpp:43]: Code that is commented out should be removed.  If needed, it can be recovered.

File: src/dfa/LG_Actions2.cpp [hunk 6]
```diff
@@ -465,7 +465,7 @@ Ret:  T_end = token.end;
 int   LG_ParseActions::START_GRM ()
 {
    int rc = 0;
-   short p = 0, a = 0;
+   short p = 0; //, a = 0;
```
Comment (thutt) [inline src/dfa/LG_Actions2.cpp:468]: This review tool has some unusual quirks.  For one, it won't allow comments on unchanged lines, as far as I can tell... so I comment here.

If 'rc' is set to 0 and unchanged, it could be removed and function made to return 'void'.  This will generate better code; the compiler will not need to set a register to 0 before the return.

If a is not needed, please remove it from the code.

File: src/dfa/LG_Generate.cpp [hunk 7]
```diff
@@ -10,7 +10,7 @@ enum {
    ts_Tr,
    ts_Tc,
    ts_N_ELEMENTS
-} type_sizes;
+}; // type_sizes;
```
Comment (thutt) [inline src/dfa/LG_Generate.cpp:13]: Is type_sizes unused?  Why remove it?

Comment (thutt) [reply src/dfa/LG_Generate.cpp:13]: Oh, wait... that's a variable definition, not a type name?

File: src/dfa/LG_Lexer.cpp [hunk 8]
```diff
@@ -56,7 +56,7 @@ int   LG_Lexer::get_token ()     // DFA LG_Lexer Engine.
       token.start = token.end;            // Set start of token.
       token.line  = line_numb;            // Set line number of token.
       token.column = col_numb + 1;        // Set column number.
-      while (next = Tm [Tr [state] + Tc [*token.end]])
+      while ((next = Tm [Tr [state] + Tc [*token.end]]))
```
Comment (thutt) [inline src/dfa/LG_Lexer.cpp:59]: Not a fan of assignments in conditionals; it's a source of errors.
But, this is original code,  and the parens are truly not needed.

File: src/lrstar/PG_Actions3.cpp [hunk 9]
```diff
@@ -503,7 +503,7 @@ int   PG_NodeActions::tail_eof_ (void* v)
 
 int   PG_NodeActions::sep_(void* v)
 {
-   Node* np = (Node*)v;
+   //Node* np = (Node*)v;
```
Comment (thutt) [inline src/lrstar/PG_Actions3.cpp:506]: This function has become vacuous.  It should be removed.

Comment (thutt) [inline src/lrstar/PG_Actions3.cpp:506]: This function is vacuous.  Could it be removed?

File: src/lrstar/PG_Actions3.cpp [hunk 10]
```diff
@@ -168,7 +168,7 @@ int   PG_NodeActions::prod_ (void* v)
 
 int   PG_NodeActions::tail_list_(void* v)
 {
-   Node* np = (Node*)v;
+   //Node* np = (Node*)v;
```
Comment (thutt) [inline src/lrstar/PG_Actions3.cpp:171]: This function has become vacuous.
Can it be removed?

File: src/lrstar/PG_Parser.cpp [hunk 11]
```diff
@@ -699,8 +699,8 @@ void  PG_Parser::init_symtab (int max_symb)
 
 void  PG_Parser::term_symtab ()
 {
-   if (symbol != NULL) delete [] symbol;
-   if (symbol != NULL) delete [] hashvec;
+   if (symbol != NULL) FREE(symbol, max_symbols);
```
Comment (thutt) [inline src/lrstar/PG_Parser.cpp:702]: Is valgrind complaining that the allocation was done with malloc() and release with 'delete'?
Can you make that explicit in the change description?

File: src/dfa/CM_Global.h [hunk 12]
```diff
@@ -229,7 +229,7 @@ class Symtab
    static short    length;          // Current length of symbol name.
    static int      cell;            // Current hash cell number.
    static int      sti;             // Current symbol-table index.
-   static uint     hashdiv;         // Hash divisor.
+   static uint32     hashdiv;         // Hash divisor.
```
Comment (thutt) [inline src/dfa/CM_Global.h:232]: Please explain why these changes are needed for mingw64.

Does that compiler produce windows executables?  (in other words, why use it?)

File: src/lrstar/PG_Generate.cpp [hunk 13]
```diff
@@ -2027,6 +2027,7 @@ static void instantiate_main(FILE *fp, const char *grammar)
            "        delete [] input_start;\n"
            "    }\n\n"
            "    fclose(output_fp);\n"
+           "    if(options.output) free(options.output);\n"
```
Comment (thutt) [inline src/lrstar/PG_Generate.cpp:2040]: Please add a new block for the free().

if (options.output) {
  free(options.output);
}

Or simply free(options.output), because free-ing a NULL pointer is ok.

File: src/lrstar/PG_Generate.cpp [hunk 14]
```diff
@@ -1935,6 +1935,13 @@ static void instantiate_getopt(FILE *fp)
            "                                 */\n"
            "    }\n"
            "}\n"
+           "\n"
+           "    static void\n"
+           "    fatal_free_opt(int rc)\n"
```
Comment (thutt) [inline src/lrstar/PG_Generate.cpp:1940]: Please mark this with the NORETURN annotation.
It's defined via including lrstar_basic_defs.h.

Also, either just free(options.output), or put it in a full block.

**Review by thutt — CHANGES_REQUESTED** [review]
Bear with me.  I'm not used to using githubs review system.  The system I normally use is either fully done in an email converstation, or all comments are part of a review.

File: make/config.mk (hunk 1 above)
Comment (mingodad) [reply make/config.mk:58]: That's true but still there is some warnings pending.

File: src/dfa/CM_Global.cpp (hunk 2 above)
Comment (mingodad) [reply src/dfa/CM_Global.cpp:358]: That's fine to me.

File: src/dfa/CM_Global.cpp (hunk 3 above)
Comment (mingodad) [reply src/dfa/CM_Global.cpp:396]: That's how the code is in the original code.

File: src/dfa/LG_Actions2.cpp (hunk 5 above)
Comment (mingodad) [reply src/dfa/LG_Actions2.cpp:43]: In fact it need a more thorough investigation before remove it because there is several commented out code that maybe is using it when uncommented.

File: src/dfa/LG_Actions2.cpp (hunk 6 above)
Comment (mingodad) [reply src/dfa/LG_Actions2.cpp:468]: This one I agree that's ok to remove.

File: src/dfa/LG_Lexer.cpp (hunk 8 above)
Comment (mingodad) [reply src/dfa/LG_Lexer.cpp:59]: Extra parenthesis is a way to tell the compiler "I know what I'm doing here don't show me a warning".

File: src/lrstar/PG_Parser.cpp (hunk 11 above)
Comment (mingodad) [reply src/lrstar/PG_Parser.cpp:702]: That's correct, mismatch allocation/deallocation.

File: src/dfa/CM_Global.h (hunk 12 above)
Comment (mingodad) [reply src/dfa/CM_Global.h:232]: Maybe I should use "unsigned int", yes it mingw compilers does generate windows executables.

File: src/dfa/LG_Lexer.cpp (hunk 8 above)
Comment (thutt) [reply src/dfa/LG_Lexer.cpp:59]: Extra parens are useful to disambiguate some expressions in C-based languages, and some compilers will suggest it.
Emacs can also use extra parens for better syntax-based formatting.
IMO, the much, much better solution for this case is to not have the assignment in the conditional expression; it doesn't generate any better code, but it harder to reason about.

File: src/lrstar/PG_Parser.cpp (hunk 11 above)
Comment (thutt) [reply src/lrstar/PG_Parser.cpp:702]: How hard is it to set up valgrind these days?
Should the tools in ./scripts be updated to support it?
